_posts/2019-03-30-numba-stencil.md

```diff
@@ -0,0 +1,325 @@
+---
+layout: post
+title: Composing Dask Array with Numba Stencils
+author: Matthew Rocklin
+tags: [dask, numba]
+theme: twitter
+---
+{% include JB/setup %}
+
+In this post we explore four array computing technologies, and how they
+work together to achieve powerful results.
+
+1.  Numba's stencil decorator to craft localized compute kernels
+2.  Numba's Just-In-Time (JIT) compiler for array computing in Python
+3.  Dask Array for parallelizing array computations across many chunks
+4.  NumPy's Generalized Universal Functions (gufuncs) to tie everything
+    together
+
+In the end we'll show how a novice developer can write a small amount of Python
+to efficiently compute localized computation on large amounts of data.  In
+particular we'll write a simple function to smooth images and apply that in
+parallel across a large stack of images.
+
+Here is the full code, we'll dive into it piece by piece below.
+
+```python
+import numba
+
+@numba.stencil
+def _smooth(x):
+    return (x[-1, -1] + x[-1, 0] + x[-1, 1] +
+            x[ 0, -1] + x[ 0, 0] + x[ 0, 1] +
+            x[ 1, -1] + x[ 1, 0] + x[ 1, 1]) // 9
+
+
+@numba.guvectorize(
+    [(numba.int8[:, :], numba.int8[:, :])],
+    '(n, m) -> (n, m)'
+)
+def smooth(x, out):
+    out[:] = _smooth(x)
+
+
+# If you want fake data
+import dask.array as da
+x = da.ones((1000000, 1000, 1000), chunks=('auto', -1, -1), dtype='int8')
+
+# If you have actual data
+import dask_image
+x = dask_image.imread.imread('/path/to/*.png')
+
+y = smooth(x)
+# dask.array<transpose, shape=(1000000, 1000, 1000), dtype=int8, chunksize=(125, 1000, 1000)>
+```
+
+Lets break this down a bit
```

Comment:
```suggestion
Note: the `smooth` function above is more commonly referred to as the 2D mean filter in the image processing community.

Lets break this down a bit
```

Reply:
@mrocklin I thought of something like this. I think it's not of uttermost importance though, if you really hate the way it looks after this, I'm fine if you just skip it.